v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/writer/SpannerWriter.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.writer;
+
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.teleport.v2.templates.RowContext;
+import java.io.Serializable;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerConfig;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerIO;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerIO.FailureMode;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerIO.Write;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerWriteResult;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Write the rows coming in from the transformer to Spanner efficiently
+ */
+public class SpannerWriter implements Serializable {
```

Comment:
The current class has simple functionality of setting up the flow and ensuring the the rows get written out - either to Spanner or the DLQ as they seem fit. Most of the common logic is in Spanner IO already.

More stuff will be added here around batching etc. in further iterations at which point i'm not convinced this will be generic. Once that goes in we can take a call if there is value in pulling this into the common flows to be used generically.